### PR TITLE
Can create project setups with "meson init"

### PR DESCRIPTION
**Pull request by jpakkane**
Need to add C++ projects as well as a unit test that tests that all setups.

### CODE REVIEW

**Comment by ignatenkobrain** [comment]
Flake8 detected 6 issues on 0c03df50597edcd91a38764936f29c6f4a661c86
Visit https://sideci.com/gh/19784232/pull_requests/2822 to review the issues.

**Comment by behlec** [comment]
This is a great idea. However, I have one question: there has been a discussion (in some issue I cannot find right now) to use `meson init builddir` instead of the current `meson builddir` syntax. I would be very much in favor of this change and would like to get rid of the current way to initialize a build dir. Rational: There are lots of commands already in use. It would for one unify the usage of meson and also help to provide better error messages. If you misspell your command you get an error message that is not that useful. E.g. `meson configur builddir`  yields

    Error during basic setup:
    
    Neither directory contains a meson.build file

If that change would be done I would see `init` als a natural candidate. Maybe `create` would be a suitable name for this feature?

**Comment by jpakkane** [comment]
I chose init because that is what other systems do (at least Cargo). We can name it something else if this seems confusing.

**Comment by ignatenkobrain** [comment]
Flake8 detected 9 issues on db9d2130aaaf83858620bf272e1cd757b61e32ad
Visit https://sideci.com/gh/19784232/pull_requests/2822 to review the issues.

**Comment by nirbheek** [comment]
I would say that `meson configure` being `mesonconf` instead of `meson` is what's confusing for people, but that's because Autoconf `./configure` is a confusing name when applied to how Meson works.

Difficult position to be in.

**Comment by behlec** [comment]
Well, my comment is based on the idea that meson should be always invoked as `meson cmd options`.
Which is how it is right now except for initializing a build directory.
This would make the interface uniform and would allow better error detection. If the fist argument is not a command meson can give a clear error message. Right now it has to assume that is meant to be a directory name. It would for example allow to do a fuzzy search on the commands and suggest the correct spelling.

Given a meson.build file one could then just do

    > meson init builddir
    > meson configure builddir
    > meson build builddir
    > meson test builddir
I this case `meson init builddir` does what currently `meson builddir` does.

This would collide with the idea of what `meson init` is doing in this pull request. I am not hung up on the idea to name this command `init`, I just wanted to point out that this might be a good usecase that should not be blocked without further thinking. Maybe `meson new` would be an idea. Then again this could be mistaken.
`meson project new` or `meson project init` could be a command to initialize a new project.

**Comment by NickeZ** [comment]
Bikeshedding ahead: I think `git init` has made it a de facto standard to initialize a project in the current dir and I think `meson init` would be the least surprising name for that command. `meson configure builddir` would make more sense as a "create build dir" in my opinion.

Following cargo we could also have `meson init --bin` for a template with an executable.

File: run_unittests.py [hunk 1]
```diff
@@ -37,8 +37,9 @@
 from mesonbuild.mesonlib import is_linux, is_windows, is_osx, is_cygwin, windows_proof_rmtree
 from mesonbuild.mesonlib import python_command, meson_command, version_compare
 from mesonbuild.environment import Environment
+from mesonbuild.environment import Environment, detect_ninja
```
Comment (ignatenkobrain) [inline run_unittests.py]: __[Flake8]__

[F811] redefinition of unused 'Environment' from line 39

File: run_unittests.py [hunk 2]
```diff
@@ -37,8 +37,9 @@
...
 from mesonbuild.mesonlib import MesonException, EnvironmentException
-from mesonbuild.dependencies import PkgConfigDependency, ExternalProgram
+from mesonbuild.dependencies import PkgConfigDependency, ExternalProgram, DependencyException
```
Comment (ignatenkobrain) [inline run_unittests.py]: __[Flake8]__

[F401] 'mesonbuild.dependencies.DependencyException' imported but unused

**Pull request merged**